### PR TITLE
Implement draft state for entity editor links table

### PR DESCRIPTION
**Pull request by yusufkinatas**
## 🌟 What is the purpose of this PR?
 This PR implements draft state for links table in entity editor. So the link entities will be created/archived when user hits `Save Changes` / `Create Entity` button. This will provide a better UX because it will works as same as properties table.

## 🔗 Related links

- [Asana task](https://app.asana.com/0/1203543021352032/1203711964051433/f) _(internal)_

## 🔍 What does this change?

- Add `useDraftLinkState` - creates draft link states
- Add `useApplyDraftLinkEntityChanges` - returns a function which applies the draft changes stored in the draft link states.
- Add `useMarkLinkEntityToArchive` - returns a function which marks the given link entity to archive. Created this to reuse the following logic: "checking if given link entity is a draft or a real entity" 

## ❓ How to test this?

1.  Checkout the branch / view the deployment
1.  Open entity editor for an entity type with links
1.  Confirm that links table works as expected while creating and editing entity (only applies the changes when user hits "save change" button, and discards the changes on "discard" button)

## 📹 Demo

Uploading Screen Recording 2023-01-16 at 15.33.16.mov…

### CODE REVIEW

File: packages/hash/frontend/src/pages/[shortname]/entities/[entity-uuid].page/entity-editor-page.tsx [hunk 1]
```diff
@@ -13,18 +13,15 @@ interface EntityEditorPageProps extends EntityEditorProps {
...
   entityLabel,
   editBar,
   entityUuid,
   owner,
-  entitySubgraph,
-  setEntity,
-  refetch,
-  hideLinksSection,
+  // `rest` is basically `EntityEditorProps` here, extracted other props above
+  ...rest
```
Comment (CiaranMn) [inline packages/hash/frontend/src/pages/[shortname]/entities/[entity-uuid].page/entity-editor-page.tsx:24]:
```suggestion
  ...entityEditorProps
```

Comment (yusufkinatas) [reply packages/hash/frontend/src/pages/[shortname]/entities/[entity-uuid].page/entity-editor-page.tsx]: Changed here https://github.com/hashintel/hash/pull/1840/commits/c9e763b6f8fc63859cb2ce8f006fa5301d67c5aa

File: packages/hash/frontend/src/pages/[shortname]/entities/[entity-uuid].page.tsx [hunk 2]
```diff
@@ -41,6 +45,20 @@ const Page: NextPageWithLayout = () => {
...
+    draftLinksToArchive,
+    setDraftLinksToArchive,
+  ] = useDraftLinkState();
+
+  useEffect(() => {
+    // set as dirty if draft link arrays are touched
+    if (draftLinksToCreate.length || draftLinksToArchive.length) {
+      setIsDirty(true);
+    }
+  }, [draftLinksToCreate, draftLinksToArchive]);
```
[the remark this answers is not in the document]
Comment (yusufkinatas) [reply packages/hash/frontend/src/pages/[shortname]/entities/[entity-uuid].page.tsx]: I totally agree, thanks for the suggestion! Applied here https://github.com/hashintel/hash/pull/1840/commits/c9e763b6f8fc63859cb2ce8f006fa5301d67c5aa

File: packages/hash/frontend/src/pages/[shortname]/entities/[entity-uuid].page/entity-editor/links-section/link-table/cells/linked-with-cell/linked-entity-list-editor.tsx [hunk 3]
```diff
@@ -56,39 +74,34 @@ export const LinkedEntityListEditor: ProvideEditorComponent<LinkedWithCell> = (
...
-      data: {
-        entityTypeId: linkEntityTypeId,
-        properties: {},
-        linkData: {
-          leftEntityId: getRoots(entitySubgraph)[0]?.metadata.editionId.baseId!,
-          rightEntityId: selectedEntity.metadata.editionId.baseId,
-        },
-      },
+    const leftEntityId = getRoots(entitySubgraph)[0]?.metadata.editionId
+      .baseId! as EntityId;
```
[the remark this answers is not in the document]
Comment (yusufkinatas) [reply packages/hash/frontend/src/pages/[shortname]/entities/[entity-uuid].page/entity-editor/links-section/link-table/cells/linked-with-cell/linked-entity-list-editor.tsx]: I guess somehow `!` at the end get forgotten there. Good catch! Removed it here https://github.com/hashintel/hash/pull/1840/commits/c9e763b6f8fc63859cb2ce8f006fa5301d67c5aa

**Review by CiaranMn — APPROVED** [review]

**Pull request merged**